feat(magic-link): add handler function support for disableSignUp

GoPro16 opened the pull request:
## Summary
Allow `disableSignUp` to accept a function that receives the endpoint context and returns a boolean. This enables conditional signup blocking based on request data, application state, or external conditions.

## Changes
- Updated `disableSignUp` option to accept `boolean | ((ctx: GenericEndpointContext) => boolean | Promise<boolean>)`
- Added support for both synchronous and asynchronous handler functions
- Function receives full endpoint context with access to `body`, `context`, and `request`
- Function is evaluated in both the send and verify endpoints
- Only queries the database when signup is actually disabled (efficient)

## Example Usage
```typescript
magicLink({
  sendMagicLink: async ({ email, url, token }) => {
    await sendEmail(email, url);
  },
  disableSignUp: (ctx) => {
    // Only allow signups from specific email domains
    const email = ctx.body.email;
    const allowedDomains = ["company.com", "partner.com"];
    const domain = email.split("@")[1];
    return !allowedDomains.includes(domain);
  }
})
```

Or with async logic:
```typescript
magicLink({
  disableSignUp: async (ctx) => {
    // Check feature flag or application state
    const signupsEnabled = await checkFeatureFlag("magic-link-signups");
    return !signupsEnabled;
  }
})
```

## Tests
- ✅ All existing tests pass
- ✅ Added comprehensive tests for function handler behavior
- ✅ Tests cover sync/async functions, context access, and edge cases
- ✅ Test verifies function is evaluated during verification flow
- ✅ TypeScript compilation successful with proper types

GoPro16 committed:
feat(magic-link): add handler function support for disableSignUp

Allow disableSignUp to accept a function that receives the request context
and returns a boolean. This enables conditional signup blocking based on
request headers, paths, or other contextual information.

The function supports both sync and async handlers and properly handles
cases where the request may be undefined.

Example usage:
magicLink({
  disableSignUp: (request) => {
    if (!request) return false;
    const referer = request.headers.get("referer");
    return referer?.includes("/admin") ?? false;
  }
})

GoPro16 committed:
fix: evaluate disableSignUp function in verification flow

The verification endpoint now properly evaluates the disableSignUp function
(if provided) instead of just checking for its existence. This ensures that
the handler's return value is respected during the magic link verification
flow, not just during the initial send.

Addresses feedback from cubic-dev-ai bot review.

GoPro16 reviewed (commented):
Sorry for the spam 😄

GoPro16 committed:
docs(magic-link): document disableSignUp function handler support

Add documentation for the new ability to pass a function to disableSignUp,
including examples for both sync and async handlers, and note that the
request parameter may be undefined in some contexts.

himself65 commented on docs/content/docs/plugins/magic-link.mdx:
```diff
@@ -134,7 +134,38 @@ and a `request` object as the second parameter.
+    sendMagicLink: async ({ email, token, url }, request) => {
+        // send email
+    },
+    disableSignUp: (request) => {
+        if (!request) return false;
+        const referer = request.headers.get("referer");
+        return referer?.includes("/admin") ?? false;
+    }
```
From logical consideration, this is incorrect. You cannot trust data from http request, so disabling based on referer is insecure

GoPro16 replied:
Agreed there, can definitely change the example. The use case we have right now is more of a UX consideration than blocking a potential attacker from spoofing the referer. ie. we want to check if the user is signing up from our sign-up page and not our sign-in page as the same `signIn.magicLink` is used for both when `disableSignUp` is `false`

GoPro16 replied:
Updated the example to check blocked domain list instead along side the context feedback.
https://github.com/better-auth/better-auth/pull/5089/commits/5a48c8195a4c8dc81db9ae095a29eb828ec588e9

Bekacru commented on packages/better-auth/src/plugins/magic-link/index.ts:
```diff
@@ -364,7 +381,17 @@ export const magicLink = (options: MagicLinkopts) => {
 
 					if (!user) {
-						if (!opts.disableSignUp) {
+						// Handle disableSignUp - can be boolean or function
+						let signUpDisabled = false;
+						if (opts.disableSignUp) {
+							if (typeof opts.disableSignUp === "function") {
+								signUpDisabled = await opts.disableSignUp(ctx.request);
+							} else {
+								signUpDisabled = true;
+							}
+						}
```
```suggestion
						const signUpDisabled = typeof opts.disableSignUp === "function" ? await opts.disableSignUp(ctx.request) : opts.disableSignUp
```

Bekacru commented on packages/better-auth/src/plugins/magic-link/index.ts:
```diff
@@ -29,9 +29,16 @@ interface MagicLinkopts {
-	disableSignUp?: boolean;
+	disableSignUp?:
+		| boolean
+		| ((request: Request | undefined) => boolean | Promise<boolean>);
```
I think you should pas `ctx` object instead. That's one of the refactors we're planning to do. 

`ctx: GenericEndpointContext`

GoPro16 committed:
Update packages/better-auth/src/plugins/magic-link/index.ts

Co-authored-by: Bereket Engida <86073083+Bekacru@users.noreply.github.com>

GoPro16 committed:
refactor(magic-link): pass ctx instead of request to disableSignUp

Address feedback from maintainers:
- Pass full endpoint context (ctx) instead of just Request
- Gives access to body, context, and request for more flexibility
- Update documentation with safer examples (domain allowlist, feature flags)
- Remove insecure referer header example as suggested by himself65
- Use proper GenericEndpointContext type in tests instead of any

gustavovalverde commented:
@GoPro16 can you update this PR? Please ping me when you've done it :)

GoPro16 commented:
@gustavovalverde I am actually going to close this PR as the original intention that I had for this was correct but implemented at the wrong level. Reason being that if we are trying to "block" domains typically we would not want to expose as a raw error of "signup is disabled" or "email is blocked" because of an attacker gaining insight easily that the policy exists — and from there enumerating which domains/emails the policy applies to. We also pivoted to using email-otp and handling similar logic in a before hook.
